openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,588 @@
+    """
+    def __init__(self):
+        # Map of xblock field name to the field's value for this block.
+        # dict {string: any picklable type}
+        self._xblock_fields = {}
+
+        # Map of transformer name to the transformer's data for this
+        # block.
+        # defaultdict {string: dict}
+        self._transformer_data = defaultdict(dict)
+
+
+class BlockStructureBlockData(BlockStructure):
+    """
+    Subclass of BlockStructure that is responsible for managing block
+    and transformer data.
+    """
+    def __init__(self, root_block_usage_key):
+        super(BlockStructureBlockData, self).__init__(root_block_usage_key)
+
+        # Map of a block's usage key to its collected data, including
+        # its xBlock fields and block-specific transformer data.
+        # defaultdict {UsageKey: _BlockData}
+        self._block_data_map = defaultdict(_BlockData)
+
+        # Map of a transformer's name to its non-block-specific data.
+        # defaultdict {string: dict}
+        self._transformer_data = defaultdict(dict)
+
+    def get_xblock_field(self, usage_key, field_name, default=None):
+        """
+        Returns the collected value of the xBlock field for the
+        requested block for the requested field_name; returns default if
+        not found.
+
+        Arguments:
+            usage_key (UsageKey) - Usage key of the block whose xBlock
+                field is requested.
+
+            field_name (string) - The name of the field that is
+                requested.
+
+            default (any type) - The value to return if a field value is
+                not found.
+        """
+        block_data = self._block_data_map.get(usage_key)
+        return block_data._xblock_fields.get(field_name, default) if block_data else default
+
+    def get_transformer_data(self, transformer, key, default=None):
+        """
+        Returns the value associated with the given key from the given
+        transformer's data dictionary; returns default if not found.
+
+        Arguments:
+            transformer (BlockStructureTransformer) - The transformer
+                whose collected data is requested.
+
+            key (string) - A dictionary key to the transformer's data
+                that is requested.
+        """
+        return self._transformer_data.get(transformer.name(), {}).get(key, default)
+
+    def set_transformer_data(self, transformer, key, value):
+        """
+        Updates the given transformer's data dictionary with the given
+        key and value.
+
+        Arguments:
+            transformer (BlockStructureTransformer) - The transformer
+                whose data is to be updated.
+
+            key (string) - A dictionary key to the transformer's data.
+
+            value (any picklable type) - The value to associate with the
+                given key for the given transformer's data.
+        """
+        self._transformer_data[transformer.name()][key] = value
```

Comment:
I want this layer to have the flexibility of changing its transformer data storage whenever and however it wants to in the future, without needing to update all its clients.  That's why we have the explicit getter and setter for the transformer data.